### PR TITLE
feat(AzureIpRange): create private dns zone and link to virtual network

### PR DESCRIPTION
**Pull request by ijovovic**
Changes proposed in this pull request:

When creating/deleting Azure IpRange, also create/delete:
- Private Dns zone (tied to CM network)
- Virtual network link - from above zone to Kyma network

### CODE REVIEW

File: pkg/kcp/provider/azure/iprange/new.go
[the remark this answers is not in the document]
Comment (ijovovic) [reply pkg/kcp/provider/azure/iprange/new.go:1]: Checks added to test.

File: pkg/kcp/provider/azure/iprange/client/client.go [hunk 1]
```diff
@@ -30,10 +33,22 @@ func NewClientProvider() azureclient.ClientProvider[Client] {
...
+		if err != nil {
+			return nil, err
+		}
+
 		return newClient(
 			azureclient.NewResourceGroupClient(resourcesClientFactory.NewResourceGroupsClient()),
 			azureclient.NewSecurityGroupsClient(networkClientFactory.NewSecurityGroupsClient()),
 			azureclient.NewSubnetsClient(networkClientFactory.NewSubnetsClient()),
+			azureclient.NewPrivateDnsZoneClient(privateDnsZoneClientFactory.NewPrivateZonesClient()),
+			azureclient.NewVirtualNetworkLInkClient(virtualNetworkLinkClientFactory.NewVirtualNetworkLinksClient()),
```
[the remark this answers is not in the document]
Comment (ijovovic) [reply pkg/kcp/provider/azure/iprange/client/client.go]: thanks :) fixed

File: pkg/kcp/provider/azure/iprange/new.go [hunk 2]
```diff
@@ -39,10 +39,15 @@ func New(stateFactory StateFactory) composed.Action {
 			namesDetermine,
 			securityGroupLoad,
 			subnetLoad,
+			privateDnsZoneLoad,
+			privateVirtualNetworkLinkLoad,
 			composed.IfElse(
 				composed.MarkedForDeletionPredicate,
 				composed.ComposeActions(
 					"azureIpRangeDelete",
+					privateVirtualNetworkLinkDelete,
```
[the remark this answers is not in the document]
Comment (ijovovic) [reply pkg/kcp/provider/azure/iprange/new.go]: Indeed. Renaming done.

File: internal/controller/cloud-control/iprange_azure_test.go [hunk 3]
```diff
@@ -205,6 +207,36 @@ var _ = Describe("Feature: KCP IpRange for Azure", func() {
 			Expect(azureSecurityGroup).NotTo(BeNil())
 		})
 
+		var virtualNetworkLink *armprivatedns.VirtualNetworkLink
```
Comment (vandjelk) [inline internal/controller/cloud-control/iprange_azure_test.go:210]: You don't need eventually here. Just move it after https://github.com/kyma-project/cloud-manager/pull/687/files#diff-c52b7ce9733d241804c6e5e720b8fcb2b519c25f189ed988c2a8cb5be97a5f06R263

File: internal/controller/cloud-control/iprange_azure_test.go [hunk 4]
```diff
@@ -205,6 +207,36 @@ var _ = Describe("Feature: KCP IpRange for Azure", func() {
...
+				vnl, err := azureMock.GetVirtualNetworkLink(infra.Ctx(), resourceGroupName, privateDnsZoneName, kcpIpRangeName)
+				if err != nil {
+					return fmt.Errorf("error loading azure Virtual Network Link: %w", err)
+				}
+				virtualNetworkLink = vnl
+				return nil
+			}).
+				Should(Succeed())
+			Expect(virtualNetworkLink).NotTo(BeNil())
+		})
```
Comment (dushanpantic) [inline internal/controller/cloud-control/iprange_azure_test.go:224]: In other places where we had a similar check, we would usually add a field to the status that would provide a reference to it.
We would then check that the field is set via ObjAssertion (e.g. `HavingStatusVirtualNetworkLink`).
When it becomes set, we would query the mock outside of the "eventually" block to fetch it.

See example here:
https://github.com/kyma-project/cloud-manager/blob/236bee374e43c2c24999065a3458b93e92b2c660/internal/controller/cloud-control/nfsinstance_aws_test.go#L88-L95

File: internal/controller/cloud-control/iprange_azure_test.go [hunk 5]
```diff
@@ -205,6 +207,36 @@ var _ = Describe("Feature: KCP IpRange for Azure", func() {
...
+				pdz, err := azureMock.GetPrivateDnsZone(infra.Ctx(), resourceGroupName, privateDnsZoneName)
+				if err != nil {
+					return fmt.Errorf("error loading azure Private Dns Zone: %w", err)
+				}
+				privateDnsZone = pdz
+				return nil
+			}).
+				Should(Succeed())
+			Expect(privateDnsZone).NotTo(BeNil())
+		})
```
Comment (dushanpantic) [inline internal/controller/cloud-control/iprange_azure_test.go:238]: Same as comment above

File: pkg/kcp/provider/azure/iprange/privateDnsZoneCreate.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	privateDnsZoneName := azureutil.NewPrivateDnsZoneName(state.ObjAsIpRange().Name)
+	privateDNSZone := armprivatedns.PrivateZone{
+		Location: to.Ptr("global"),
+	}
+
+	err := state.azureClient.CreatePrivateDnsZone(ctx, state.resourceGroupName, privateDnsZoneName, privateDNSZone)
+
+	if azuremeta.IsTooManyRequests(err) {
+		return azuremeta.LogErrorAndReturn(err, "Azure KCP IpRange too many requests on create privateDnsZone", ctx)
+	}
```
Comment (dushanpantic) [inline pkg/kcp/provider/azure/iprange/privateDnsZoneCreate.go:34]: Would it make sense to requeue with delay here?

File: pkg/kcp/provider/azure/iprange/privateDnsZoneDelete.go [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+	logger.Info("Deleting Azure KCP IpRange dnsZone")
+
+	resourceGroupName := state.resourceGroupName
+	privateDnsZoneName := state.privateDnsZone.Name
+
+	err := state.azureClient.DeletePrivateDnsZone(ctx, resourceGroupName, ptr.Deref(privateDnsZoneName, ""))
+	if azuremeta.IsTooManyRequests(err) {
+		return azuremeta.LogErrorAndReturn(err, "Azure KCP IpRange too many requests on delete dnsZone", ctx)
+	}
```
Comment (dushanpantic) [inline pkg/kcp/provider/azure/iprange/privateDnsZoneDelete.go:29]: Same as previous comment, perhaps requeue with delay, if it makes sense?

File: pkg/kcp/provider/azure/iprange/virtualNetworkLinkCreate.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+					state.Scope().Spec.Scope.Azure.VpcNetwork, kymaNetworkName).String()),
+			},
+			RegistrationEnabled: ptr.To(false),
+		},
+	}
+	err := state.azureClient.CreateVirtualNetworkLink(ctx, resourceGroupName, privateDnsZoneName, virtualNetworkLinkName, virtualNetworkLink)
+
+	if azuremeta.IsTooManyRequests(err) {
+		return azuremeta.LogErrorAndReturn(err, "Azure KCP IpRange too many requests on create virtualNetworkLink", ctx)
+	}
```
Comment (dushanpantic) [inline pkg/kcp/provider/azure/iprange/virtualNetworkLinkCreate.go:43]: Same as previous comment, perhaps requeue with delay, if it makes sense?

File: pkg/kcp/provider/azure/iprange/virtualNetworkLinkDelete.go [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+	logger.Info("Deleting Azure KCP IpRange virtualNetworkLink")
+
+	resourceGroupName := state.resourceGroupName
+	privateZoneName := state.privateDnsZone.Name
+	virtualNetworkLinkName := state.virtualNetworkLink.Name
+
+	err := state.azureClient.DeleteVirtualNetworkLink(ctx, resourceGroupName, ptr.Deref(privateZoneName, ""), ptr.Deref(virtualNetworkLinkName, ""))
+	if azuremeta.IsTooManyRequests(err) {
+		return azuremeta.LogErrorAndReturn(err, "Azure KCP IpRange too many requests on delete virtualNetworkLink", ctx)
+	}
```
Comment (dushanpantic) [inline pkg/kcp/provider/azure/iprange/virtualNetworkLinkDelete.go:30]: Same as previous comment, perhaps requeue with delay, if it makes sense?

File: internal/controller/cloud-control/iprange_azure_test.go (hunk 4 above)
Comment (ijovovic) [reply internal/controller/cloud-control/iprange_azure_test.go:224]: No, in the case of Azure we know the reference to dnsZone and virtualPrivateLink based on "well known name" pattern (UUID).

We do not add this ref to KCP CR (as we do for the AWS or CCEE), so I can not use proposed pattern.

File: internal/controller/cloud-control/iprange_azure_test.go (hunk 5 above)
Comment (ijovovic) [reply internal/controller/cloud-control/iprange_azure_test.go:238]: No, in the case of Azure we know the reference to dnsZone and virtualPrivateLink based on "well known name" pattern (UUID).

We do not add this ref to KCP CR (as we do for the AWS or CCEE), so I can not use proposed pattern.

**Review by dushanpantic — APPROVED** [review]

**Pull request merged**